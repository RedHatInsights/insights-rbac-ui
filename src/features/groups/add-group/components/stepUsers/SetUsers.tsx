import React, { useCallback, useEffect, useState } from 'react';
import { useFlag } from '@unleash/proxy-client-react';
import { Stack } from '@patternfly/react-core';
import { StackItem } from '@patternfly/react-core';
import useFieldApi from '@data-driven-forms/react-form-renderer/use-field-api';
import useFormApi from '@data-driven-forms/react-form-renderer/use-form-api';
import { UsersList } from './UsersList';
import { ActiveUsers } from '../../../../../components/user-management/ActiveUsers';
import type { User } from './types';
import '../../../../../App.scss';

interface SetUsersProps {
  name: string;
  // Data-driven-forms props
  input?: any;
  meta?: any;
  [key: string]: any;
}

export const SetUsers: React.FC<SetUsersProps> = (props) => {
  const formOptions = useFormApi();
  const [selectedUsers, setSelectedUsers] = useState<User[]>(formOptions.getState().values['users-list'] || []);
  const { input } = useFieldApi(props);
  const isITLess = useFlag('platform.rbac.itless');

  useEffect(() => {
    input.onChange(selectedUsers);
    formOptions.change('users-list', selectedUsers);
  }, [selectedUsers]); // Remove unstable formOptions and input dependencies

  // Handle selection changes from UsersList - sync with form API
  const handleUserSelection = useCallback((users: User[]) => {
    setSelectedUsers(users);
  }, []);

  const activeUserProps = {
    linkDescription: 'Select users from your user access list to add to this group.',
  };

  return (
    <Stack hasGutter>
      <StackItem>
        {isITLess ? (
          <UsersList initialSelectedUsers={selectedUsers} onSelect={handleUserSelection} displayNarrow={true} />
        ) : (
          <ActiveUsers {...activeUserProps}>
            <UsersList initialSelectedUsers={selectedUsers} onSelect={handleUserSelection} displayNarrow={true} />
          </ActiveUsers>
        )}
      </StackItem>
    </Stack>
  );
};
