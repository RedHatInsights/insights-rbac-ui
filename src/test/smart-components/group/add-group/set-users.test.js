import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';
import SetUsers from '../../../../smart-components/group/add-group/set-users';

let initialProps;
let initialState;
const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
let mockStore;

describe('<SetUsers />', () => {
  beforeEach(() => {
    initialState = {
      roleReducer: {
        roles: [
          {
            label: 'foo',
            value: 'bar',
          },
        ],
      },
      groupReducer: {
        groups: {
          data: [
            {
              uuid: '123',
              name: 'SampleGroup',
            },
          ],
        },
      },
      userReducer: {
        selectedUser: {},
        isUserDataLoading: false,
        users: [],
      },
    };
    mockStore = configureStore(middlewares);
    initialProps = {
      setGroupData: jest.fn(),
      selectedUsers: [],
      setSelectedUsers: jest.fn(),
      optionIdx: 0,
      setOptionIdx: jest.fn(),
      createOption: jest.fn(),
    };
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/groups/', '/groups/add-group/', '/groups/']} initialIndex={1}>
          <SetUsers {...initialProps} />
        </MemoryRouter>
      </Provider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
