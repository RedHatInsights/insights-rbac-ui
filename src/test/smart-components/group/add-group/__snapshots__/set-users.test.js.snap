// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SetUsers /> should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "/groups/",
        "/groups/add-group/",
        "/groups/",
      ]
    }
    initialIndex={1}
  >
    <SetUsers
      createOption={[MockFunction]}
      optionIdx={0}
      selectedUsers={Array []}
      setGroupData={[MockFunction]}
      setOptionIdx={[MockFunction]}
      setSelectedUsers={[MockFunction]}
    />
  </MemoryRouter>
</ContextProvider>
`;
